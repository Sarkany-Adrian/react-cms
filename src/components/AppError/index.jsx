// @flow
import React from 'react';

function AppError(): React$Element<'div'> {
  return <div>something went wrong</div>;
}

export default AppError;
