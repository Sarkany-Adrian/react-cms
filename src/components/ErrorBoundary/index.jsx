// @flow
import * as React from 'react';

type FallbackComponentProps = {
  error: Error,
  info: Object
};

type ErrorBoundaryProps = {
  children: React.Node,
  FallbackComponent: React.ComponentType<FallbackComponentProps>,
  onError: ?(error: Error, info: Object) => void
};

type ErrorBoundaryState = {
  error: ?Error,
  info: ?Object
};

class ErrorBoundary extends React.Component<
  ErrorBoundaryProps,
  ErrorBoundaryState
> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { error: null, info: null };
  }

  componentDidCatch(error: Error, info: Object) {
    const { onError } = this.props;
    // Catch errors in any components below and re-render with error message
    this.setState({
      error,
      info
    });

    // Callback for error messages
    if (onError) {
      onError(error, info);
    }
  }

  render() {
    const { FallbackComponent, children } = this.props;
    const { error, info } = this.state;

    if (error && info) {
      // If an error occurs, render the error component
      return <FallbackComponent error={error} info={info} />;
      // return React.createElement(fallbackComponent, { error, info });
    }

    // Or just render children
    return children;
  }
}

export default ErrorBoundary;
