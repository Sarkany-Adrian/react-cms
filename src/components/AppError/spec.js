import React from 'react';
import { shallow } from 'enzyme';

import AppError from 'components/AppError';

describe('Error component', () => {
  it('should render the error component', () => {
    const wrapper = shallow(<AppError />);
    expect(
      wrapper.equals(<div>Something went wrong, please try again later.</div>)
    );
  });
});
