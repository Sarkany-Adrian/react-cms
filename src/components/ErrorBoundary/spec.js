import React from 'react';
import { shallow } from 'enzyme';

import ErrorBoundary from 'components/ErrorBoundary';
import AppError from 'components/AppError';
import toJson from 'enzyme-to-json';

describe('Error Boundary component', () => {
  it('should render children - no error', () => {
    const defaultProps = {
      children: <span>test children</span>,
      FallbackComponent: AppError,
      onError: jest.fn()
    };

    const wrapper = shallow(<ErrorBoundary {...defaultProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render fallback component - with error', () => {
    const defaultProps = {
      children: <span>test children</span>,
      FallbackComponent: AppError,
      onError: jest.fn()
    };

    const wrapper = shallow(<ErrorBoundary {...defaultProps} />);
    wrapper.setState({
      error: true,
      info: 'test info'
    });

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
