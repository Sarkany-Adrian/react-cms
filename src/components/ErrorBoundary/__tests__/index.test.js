import React from 'react';
import { shallow, mount } from 'enzyme';

import ErrorBoundary from 'components/ErrorBoundary';
import AppError from 'components/AppError';
import toJson from 'enzyme-to-json';

const ProblemChild = () => null;

/**
 *
 */
test('should render children - no error', () => {
  const defaultProps = {
    children: <span>test children</span>,
    FallbackComponent: AppError,
    onError: jest.fn()
  };

  const wrapper = shallow(<ErrorBoundary {...defaultProps} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

/**
 *
 */
test('should render fallback component - with error', () => {
  const defaultProps = {
    children: <span>test children</span>,
    FallbackComponent: AppError,
    onError: jest.fn()
  };

  const wrapper = shallow(<ErrorBoundary {...defaultProps} />);
  wrapper.setState({
    error: true,
    info: 'test info'
  });

  expect(toJson(wrapper)).toMatchSnapshot();
});

/**
 *
 */
test('triggers component did catch lifecycle method when an error is received', () => {
  const defaultProps = {
    children: <span>test children</span>,
    FallbackComponent: AppError
  };

  const wrapper = mount(
    <ErrorBoundary {...defaultProps}>
      <ProblemChild />
    </ErrorBoundary>
  );

  const error = new Error('hi!');
  wrapper.find(ProblemChild).simulateError(error);
  expect(wrapper.state('error')).toEqual(error);
});

/**
 *
 */
test('should invoke props.onError method.', () => {
  const defaultProps = {
    children: <span>test children</span>,
    FallbackComponent: AppError,
    onError: jest.fn()
  };

  const wrapper = mount(
    <ErrorBoundary {...defaultProps}>
      <ProblemChild />
    </ErrorBoundary>
  );

  const error = new Error('hi!');
  wrapper.find(ProblemChild).simulateError(error);
  expect(defaultProps.onError).toHaveBeenCalled();
});
