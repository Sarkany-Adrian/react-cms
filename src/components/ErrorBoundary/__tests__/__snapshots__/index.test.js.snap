// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render children - no error 1`] = `
<span>
  test children
</span>
`;

exports[`should render fallback component - with error 1`] = `
<AppError
  error={true}
  info="test info"
/>
`;
